support class cache for ByteBuddy (#4858)
apache/skywalking: docs/en/FAQ/Compatible-with-other-javaagent-bytecode-processing.md

## Compatible with other javaagent bytecode processing

### Problem
1. When use skywalking agent, some other agent, such as Arthas, can't work well 
https://github.com/apache/skywalking/pull/4858

> 2. Java agent retransforming class fails with Skywalking agent, such as in this [demo](https://github.com/hengyunabc/skywalking-error-demo)

Reviewer: Waiting for update at https://github.com/SkyAPMTest/retransform-conflict-demo

Author: The demo project is submitted.

## Compatible with other javaagent bytecode processing

### Problem

> 1. when use skywalking agent, arthas can't work well 

Reviewer: ```suggestion
1. When use skywalking agent, some other agent, such as Arthas, can't work well 
```

https://github.com/apache/skywalking/pull/4858

> 2. Custom java agent retransform class fails with Skywalking agent  
>  https://github.com/hengyunabc/skywalking-error-demo

Reviewer: ```suggestion
2. Java agent retransforming class fails with Skywalking agent, such as in this [demo](https://github.com/hengyunabc/skywalking-error-demo)
```

Also, this code, would you like to push them into [SkyAPMTest](https://github.com/SkyAPMTest), I could create a repo for you and invite you in. The reason I recommend there is, we host unofficial tests and demo codes there, which need to be linked from the official repo.

Author: Yes, please create a repo: skywalking-retransform-demo

Reviewer: Repo  https://github.com/SkyAPMTest/retransform-conflict-demo created, and sent the invitation to you.

 
### Reason

> Skywalking agent uses ByteBuddy to retransform classes when java application starts. 

Reviewer: ```suggestion
SkyWalking agent uses ByteBuddy to transform classes when the Java application starts. 
```

ByteBuddy generates auxiliary classes with different random names every time. 

> When other javaagent retransforms the same class, it will trigger the SkyWalking agent to enhance the class again. 

Reviewer: ```suggestion
When another java agent retransforms the same class, it triggers the SkyWalking agent to enhance the class again. 
```

> The bytecode regenerated by ByteBuddy is changed, the fields and imports are modified, the jvm verifys class bytecode failed, and the retransform fails.

Reviewer: ```suggestion
The bytecode regenerated by ByteBuddy is changed, the fields are modified, the JVM verifications about class bytecode fail, causing the retransform fails.
```

I think there is no `imports` in the byte codes level. All classes are in the full name.

Author: The `imports ` will be converted to constant table items. ByteBuddy create random name auxiliary class, and add import item to origin class.

Reviewer: Update the expression.

### Resolve

**1.Enable the class cache feature**  

Add JVM parameters:  
`-Dskywalking.agent.is_cache_enhanced_class=true -Dskywalking.agent.class_cache_mode=MEMORY`    

Or uncomment options in `agent.conf`:
  
```
# If true, SkyWalking agent will cache all instrumented classes files to memory or disk files (decided by class cache mode),
# allow other javaagent to enhance those classes that enhanced by SkyWalking agent.
agent.is_cache_enhanced_class = ${SW_AGENT_CACHE_CLASS:false}

# The instrumented classes cache mode: MEMORY or FILE
# MEMORY: cache class bytes to memory, if instrumented classes is too many or too large, it may take up more memory
# FILE: cache class bytes to user temp folder starts with 'class-cache', automatically clean up cached class files when the application exits
agent.class_cache_mode = ${SW_AGENT_CLASS_CACHE_MODE:MEMORY}

```

> If the class cahce feature is enabled, save the class bytecode of the ByteBuddy retransform to memory or a temporary file. 

Reviewer: ```suggestion
If the class cache feature is enabled, save the instrumented class bytecode to memory or a temporary file. 
```

> When other java agents retransform the same class, Skywalking agent first attempts to load from the cache.

Reviewer: ```suggestion
When other java agents retransform the same class, SkyWalking agent first attempts to load from the cache.
```

If the cached class is found, it will be used directly without regenerating a new random name auxiliary class, 
which will not affect the processing of the subsequent java agent.

**2.Class cache save mode**  

> It is recommended to put the cache class in memory, but if it takes up more memory, it should be placed in a temporary file. Set the class cache save mode through the following options:  

Reviewer: ```suggestion
It is recommended to put the cache class in memory, meanwhile if it costs more memory resources. Another option is using the local file system. Set the class cache mode through the following options:  
```

`-Dskywalking.agent.class_cache_mode=MEMORY` : save cache classes to java memory.    

> `-Dskywalking.agent.class_cache_mode=FILE` : save cache classes to Skywalking agent path '/class-cache'.  

Reviewer: ```suggestion
`-Dskywalking.agent.class_cache_mode=FILE` : save cache classes to SkyWalking agent path '/class-cache'.  
```
